src/mon/AuthMonitor.cc

```diff
@@ -1929,6 +1876,145 @@ bool AuthMonitor::prepare_command(MonOpRequestRef op)
   return false;
 }
 
+/* Generate the caps that should be present in the entity's auth keyring
+ * after running the "fs authorize" command. This is done by merging the
+ * caps already present in the client's auth keyring with the new caps
+ * provided by the user at "fs authorize" command.
+ */
+bool AuthMonitor::_gen_wanted_caps(EntityAuth& e_auth,
+  vector<string>& newcaps, ostream* out)
+{
+  bool all_caps_same = true; // whether all caps passed are same as the
+			     // one present.
+
+  if (e_auth.caps.empty()) {
+    all_caps_same = false;
+    return all_caps_same;
+  }
+
+  for (auto it = newcaps.begin(); it != newcaps.end(); it += 2) {
+    string& cap_entity = *it;
+    string& n_cap_str = *(it + 1); // new cap to be added to the current cap
+
+    if (e_auth.caps.count(cap_entity) == 0) {
+      all_caps_same = false;
+      continue;
+    }
+
+    string c_cap_str; // current cap held by entity's auth keyring
+    auto i = e_auth.caps[cap_entity].cbegin();
+    decode(c_cap_str, i);
+    if (c_cap_str == n_cap_str) {
+      continue;
+    }
+
+    if (cap_entity == "mon") {
+      MonCap c_cap, n_cap;
+      if (not c_cap.parse(c_cap_str, out))
+	*out << "error parsing " << cap_entity << "caps client already holds";
+      if (not n_cap.parse(n_cap_str, out))
+	*out << "error parsing new " << cap_entity << "caps";
+
+      if (c_cap.merge(n_cap)) {
+	all_caps_same = false;
+	n_cap_str = c_cap.str();
+      } else {
+	n_cap_str = c_cap_str;
+      }
+    } else if (cap_entity == "osd") {
+      OSDCap c_cap, n_cap;
+      if (not c_cap.parse(c_cap_str, out))
+	*out << "error parsing " << cap_entity << "caps client already holds";
+      if (not n_cap.parse(n_cap_str, out))
+	*out << "error parsing new " << cap_entity << "caps";
+
+      if (c_cap.merge(n_cap)) {
+	all_caps_same = false;
+	n_cap_str = c_cap.str();
+      } else {
+	n_cap_str = c_cap_str;
+      }
+    } else if (cap_entity == "mds") {
+      MDSAuthCaps c_cap, n_cap;
+      if (not c_cap.parse(g_ceph_context, c_cap_str, out))
+	*out << "error parsing " << cap_entity << "caps client already holds";
+      if (not n_cap.parse(g_ceph_context, n_cap_str, out))
+	*out << "error parsing new " << cap_entity << "caps";
+
+      if (c_cap.merge(n_cap)) {
+	all_caps_same = false;
+	n_cap_str = c_cap.str();
+      } else {
+	n_cap_str = c_cap_str;
+      }
```

Comment:
This code duplication can be done away with by having the respective subsystem auth caps derive from an abstract class and then calling the respective `parse` and `merge` virtual functions. WDYT?

Reply:
True. Easier thing to do is not pass g_ceph_context as it is globally available. Then the interface of all three cap classes would be the same and all these repetition will be not needed. 

Here's the PR for this - https://github.com/ceph/ceph/pull/50312. I talked about it in our last meeting and requested a review for it.

Reply:
> True. Easier thing to do is not pass g_ceph_context as it is globally available. Then the interface of all three cap classes would be the same and all these repetition will be not needed.

Could you elaborate a bit more. How does not passing g_ceph_context help with parse/merge for auth caps?

> 
> Here's the PR for this - #50312. I talked about it in our last meeting and requested a review for it.

Reply:
The interfaces for all 3 would be the same, same call would work for all three.

Reply:
In that case, we should take this approach. And, I think we could invoke the relevant routine using typecasting.

Reply:
Yes. Once that PR is merged, I'll change this code accordingly. I was expecting it too happen sooner but reviews as usual are taking time.

Reply:
> Yes. Once that PR is merged, I'll change this code accordingly. I was expecting it too happen sooner but reviews as usual are taking time.

I added `needs-qa` to that PR.